torch/onnx/_internal/fx/passes/functionalization.py

```diff
@@ -0,0 +1,79 @@
+from __future__ import annotations
+
+import torch
+import torch.func
+import torch.fx
+
+from torch.fx.experimental import proxy_tensor
+from torch.onnx._internal import _beartype
+from torch.onnx._internal.fx import _pass
+from torch.onnx._internal.fx.passes import utils
+
+
+class Functionalize(_pass.Transform):
+    """Functionalize a GraphModule.
+
+    This pass utilizes ``torch.func.functionalize`` to convert a GraphModule into a
+    functional form. The two main functionalities are:
+
+    * ``torch.func.functionalize`` removes (intermediate) mutations and aliasing from a
+    function, while preserving the function's semantics.
+
+    * ``torch.func.functionalize`` also removes mutations (and views) that were performed
+    on function inputs. However to preserve semantics, functionalize will "fix up" the
+    mutations after the transform has finished running, by detecting if any tensor inputs
+    "should have" been mutated, and copying the new data back to the inputs if necessary.
```

Comment:
How does `copying the new data back to the inputs` mean?

Reply:
Consider
```python
def fn(a, b):
    a.add_(b)
    return a
```
When you call this function `fn(x, y)`, `x` is also mutated. So just functionalizing is not enough for preserving the original semantics. A "special" input mutation step, if needed, needs to be inserted at the end. That's what the comment means. 
```python
# after functionalization
# this is not semantically the same. `a`'s value is not updated.
def fn(a, b):
    new_a = a + b
    return new_a

# functionalization that preserves semantics
def fn(a, b):
    new_a = a + b

    # copying the new data back to the inputs
    a.copy_(new_a)

    return new_a
```

Reply:
added above into docstring